### PR TITLE
docs: add the-internal-of-apisix-java-plugin-runner.md, development.md and how-it-works.md

### PR DESCRIPTION
**Pull request by tzssangglass**
add docs

### CODE REVIEW

**Comment by tzssangglass** [comment]
@spacewander @dmsolr @nic-chen pls review

File: docs/development.md [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+ ./mvnw test
+```
+
+
+### Mimic practical environment
+
+If you want to mimic the practical environment, you need to configure the route on APISIX
+by having the request go through the filter you want to test, for example
+
+```yaml
```
Comment (spacewander) [inline docs/development.md:107]: Should be json

File: docs/how-it-works.md [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+```shell
+java -jar -DAPISIX_LISTEN_ADDRESS=/tmp/runner.sock -DAPISIX_CONF_EXPIRE_TIME=3600 /path/to/apisix-java-plugin-runner.jar
+```
+
+and add the following configure in the `config.yaml` file of APISIX
+
+```yaml
+ext-plugin:
+  path_for_test: /tmp/runner.sock
+  cmd: ['echo', 'call runner']
```
Comment (spacewander) [inline docs/how-it-works.md:31]: We don't need to configure the `cmd` section as we don't need it to work.

File: docs/how-it-works.md [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+```
+
+The `/tmp/runner.sock` is the address of the file where apisix-java-plugin-runner  
+and APISIX communicate between processes and must be consistent.
+
+Note: If you see some error logs like
+
+```
+phase_func(): failed to connect to the unix socket unix:/tmp/runner.sock: permission denied
```
Comment (spacewander) [inline docs/how-it-works.md:40]: Why there is permission denied when the sock file is created by runner itself?

File: docs/development.md [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+            {
+                "name":"FooFilter",
+                "value":"bar"
+            }
+        ]
+    }
+}
+```
+
+and then by making a request to APISIX to trigger the route.
```
Comment (spacewander) [inline docs/development.md:120]: `and then make a request `

File: docs/how-it-works.md [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+and before start, you need to set the following two environment variables:
+
+- APISIX_LISTEN_ADDRESS: apisix-java-plugin-runner and APISIX for inter-process communication (Unix Domain Socket) socket type file address.
+  And do not need to actively create this file, apisix-java-plugin-runner will automatically create this file when it starts.
+- APISIX_CONF_EXPIRE_TIME: the time that APISIX's configuration is cached in the apisix-java-plugin-runner process.
+
+For example, if you start apisix-java-plugin-runner as a jar package, pass the environment variables as follows
+
+```shell
+java -jar -DAPISIX_LISTEN_ADDRESS=/tmp/runner.sock -DAPISIX_CONF_EXPIRE_TIME=3600 /path/to/apisix-java-plugin-runner.jar
```
Comment (spacewander) [inline docs/how-it-works.md:23]: The environment variable APISIX_LISTEN_ADDRESS set by APISIX is "unix:/tmp/runner.sock", see https://github.com/apache/apisix/blob/09e298302cbc8514d8240c300974746c9f85845c/apisix/plugins/ext-plugin/helper.lua#L45

File: docs/the-internal-of-apisix-java-plugin-runner.md [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+## Table of Contents
+
+- [Overview](#overview)
+- [Communication](#communication)
+- [Serialization](#serialization)
+- [Codec](#codec)
+
+## Overview
+
+The apisix-java-plugin-runner designed as a tcp server built using [reactor-netty](https://github.com/reactor/reactor-netty),
```
Comment (spacewander) [inline docs/the-internal-of-apisix-java-plugin-runner.md:14]: tcp => TCP

File: docs/the-internal-of-apisix-java-plugin-runner.md [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+Users only need to focus on their business logic, not on the details of how the apisix java plugin runner communicates with APISIX.
+
+The inter-process communication between them is depicted by the following diagram.
+
+![the-internal-of-apisix-java-plugin-runner](./images/the-internal-of-apisix-java-plugin-runner.png)
+
+## Communication
+
+apisix-java-plugin-runner and APISIX use the Unix Domain Socket for inter-process communication,
+so they need to be deployed in the same runtime environment.
```
Comment (spacewander) [inline docs/the-internal-of-apisix-java-plugin-runner.md:26]: the same runtime environment => the same instance

File: docs/the-internal-of-apisix-java-plugin-runner.md [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+apisix-java-plugin-runner is managed by APISIX. APISIX starts the apisix-java-plugin-runner when it starts and ends it when it
+ends. if the apisix-java-plugin-runner quits in the middle, APISIX will restart it automatically.
+
+## Serialization
+
+Refer to [flatbuffers](https://github.com/google/flatbuffers)
+
+FlatBuffers is a cross platform serialization library architected for maximum memory efficiency.
+It allows you to directly access serialized data without parsing/unpacking it first, while still having great forwards/backwards compatibility.
```
Comment (spacewander) [inline docs/the-internal-of-apisix-java-plugin-runner.md:36]: forward/backward compatibility

File: docs/the-internal-of-apisix-java-plugin-runner.md [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+You can refer to the [ext-plugin.fbs](https://github.com/api7/ext-plugin-proto/blob/main/ext-plugin.fbs)
+ schema file to see how lua and java lay out the serialized objects.
```
Comment (spacewander) [inline docs/the-internal-of-apisix-java-plugin-runner.md:39]: how lua and java lay out => how Lua and Java layout

File: docs/the-internal-of-apisix-java-plugin-runner.md [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+## Codec
+
+apisix-java-plugin-runner and APISIX use a private binary protocol for coding and decoding.
+The protocol format is
+
+```
+1 byte of type + 3 bytes of length + data
+```
+
+The type can be 0 ~ 7, and the length can be [0, 8M). data length is determined by length.
```
Comment (spacewander) [inline docs/the-internal-of-apisix-java-plugin-runner.md:50]: data length is => The length of data is

File: docs/how-it-works.md (hunk 3 above)
Comment (tzssangglass) [reply docs/how-it-works.md:40]: runner has permission to the sock file, but APISIX does not, the permission denied log will appear in the APISIX error.log

File: docs/how-it-works.md (hunk 5 above)
Comment (tzssangglass) [reply docs/how-it-works.md:23]: got, "unix:/tmp/runner.sock" and "/tmp/runner.sock" are compatible

File: docs/development.md [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+./mvnw install
+```
+
+Write Filter
+------------
+
+Refer to the code in the [sample](https://github.com/apache/apisix-java-plugin-runner/tree/main/sample)
+to learn how to extend `PluginFilter`, define the order, rewrite requests and stop requests.
+
+####  Code Location
```
Comment (tzssangglass) [inline docs/development.md:27]: @dmsolr pls review here
the plugin code written by the user is supposed to be placed in the https://github.com/apache/apisix-java-plugin-runner/tree/main/runner-plugin?

**Comment by tzssangglass** [comment]
update

File: docs/development.md [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+and the name of each filter's implementation class is the return value of its overridden function `public String name()`.
+
+
+####  Rewrite Request
+
+If you perform the following function call in the filter chain of the implementation class
+
+*  request.setPath()
+*  request.setHeader()
+*  request.setArg()
```
Comment (spacewander) [inline docs/development.md:74]: How to Get/Del the headers/arguments?
Is there a javadoc to list each methods?

File: docs/development.md [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+
+```shell
+cd /path/to/apisix-java-plugin-runner
+./mvnw install
+```
+
+Write Filter
+------------
+
+Refer to the code in the [sample](https://github.com/apache/apisix-java-plugin-runner/tree/main/sample)
```
Comment (spacewander) [inline docs/development.md:24]: The sample changes data with hardcoded values. Better to show how can a plugin access the configuration.

File: docs/development.md (hunk 12 above)
Comment (tzssangglass) [reply docs/development.md:74]: update, here https://github.com/apache/apisix-java-plugin-runner/pull/19/files#diff-a09a8ee05234dc648f35846a4ddc8f94da78c228642c6226431d51540c092353R243-R265

File: docs/development.md (hunk 13 above)
Comment (tzssangglass) [reply docs/development.md:24]: update, here https://github.com/apache/apisix-java-plugin-runner/pull/19/files#diff-ed24963c66d69025a9cef9bd7407fc07a29e397a71be4602c96d6a03f8d42274R63-R78

File: docs/how-it-works.md [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+and before start, you need to set the following two environment variables:
+
+- APISIX_LISTEN_ADDRESS: apisix-java-plugin-runner and APISIX for inter-process communication (Unix Domain Socket) socket type file address.
+  And do not need to actively create this file, apisix-java-plugin-runner will automatically create this file when it starts.
+- APISIX_CONF_EXPIRE_TIME: the time that APISIX's configuration is cached in the apisix-java-plugin-runner process.
+
+For example, if you start apisix-java-plugin-runner as a jar package, pass the environment variables as follows
+
+```shell
+java -jar -DAPISIX_LISTEN_ADDRESS=unix:/tmp/runner.sock -DAPISIX_CONF_EXPIRE_TIME=3600 /path/to/apisix-java-plugin-runner.jar
```
[the remark this answers is not in the document]
Comment (tzssangglass) [reply docs/how-it-works.md:23]: add refer.

File: docs/how-it-works.md [hunk 15]
```diff
@@ -61,7 +63,7 @@ cd dist
 tar -zxvf apache-apisix-runner-bin.tar.gz
 ```
 
-then the files in the `dist` directory are laid out as follows
+file's layout in the `dist` directory are as follows
```
Comment (spacewander) [inline docs/how-it-works.md:66]: bad grammar

File: docs/how-it-works.md [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+
+If you are developing a new plugin and need to debug the code, then you can run the main class
+[PluginRunnerApplication](https://github.com/apache/apisix-java-plugin-runner/blob/main/runner-starter/src/main/java/org/apache/apisix/plugin/runner/PluginRunnerApplication.java),
+and before start, you need to set the following two environment variables:
+
+- APISIX_LISTEN_ADDRESS: apisix-java-plugin-runner and APISIX for inter-process communication (Unix Domain Socket) socket type file address.
+  And do not need to actively create this file, apisix-java-plugin-runner will automatically create this file when it starts.
+- APISIX_CONF_EXPIRE_TIME: the time that APISIX's configuration is cached in the apisix-java-plugin-runner process.
+
+For example, if you start apisix-java-plugin-runner as a jar package, pass the environment variables as follows
```
Comment (spacewander) [inline docs/how-it-works.md:20]: as follows => as below

**Review by spacewander — COMMENTED** [review]
My suggestion is that you can write doc in Chinese and translate it to English via translater.

File: docs/how-it-works.md (hunk 15 above)
Comment (tzssangglass) [reply docs/how-it-works.md]: update

File: docs/how-it-works.md (hunk 16 above)
Comment (tzssangglass) [reply docs/how-it-works.md]: update, and changed all the `as follow` to `as below`

**Review by spacewander — APPROVED** [review]

**Pull request merged**